lock-api/src/test/java/com/palantir/lock/client/ProfilingTimelockServiceTest.java

```diff
@@ -0,0 +1,231 @@
+
+    private final ProfilingTimelockService profilingTimelockService = new ProfilingTimelockService(
+            logger, delegate, () -> Stopwatch.createStarted(ticker), loggingPermissionSupplier);
+
+    @Test
+    public void doesNotLogIfOperationsAreFast() {
+        flushLogsWithCall(SHORT_DURATION, profilingTimelockService::getFreshTimestamp);
+        flushLogsWithCall(SHORT_DURATION, profilingTimelockService::startIdentifiedAtlasDbTransaction);
+
+        verify(delegate).getFreshTimestamp();
+        verify(delegate).startIdentifiedAtlasDbTransaction();
+        verifyLoggerNeverInvoked();
+    }
+
+    @Test
+    public void logsIfOperationsAreSlowAndLoggingAllowed() {
+        flushLogsWithCall(LONG_DURATION, profilingTimelockService::getFreshTimestamp);
+
+        verify(delegate).getFreshTimestamp();
+        verifyLoggerInvokedOnceWithSpecificProfile("getFreshTimestamp", LONG_DURATION);
+    }
+
+    @Test
+    public void doesNotLogIfLoggingNotAllowedEvenIfOperationsAreSlow() {
+        accumulateLogsWithCall(LONG_DURATION, profilingTimelockService::getFreshTimestamp);
+
+        verify(delegate).getFreshTimestamp();
+        verifyLoggerNeverInvoked();
+    }
+
+    @Test
+    public void returnsResultReturnedByDelegate() {
+        when(profilingTimelockService.getFreshTimestamp()).thenReturn(42L);
+        assertThat(profilingTimelockService.getFreshTimestamp()).isEqualTo(42L);
+    }
+
+    @Test
+    public void logsIfFailedOperationsAreSlow() {
+        RuntimeException exception = new RuntimeException("kaboom");
+        when(profilingTimelockService.getFreshTimestamp()).thenThrow(exception);
+
+        assertThatThrownBy(() -> flushLogsWithCall(LONG_DURATION, profilingTimelockService::getFreshTimestamp))
+                .isEqualTo(exception);
+
+        verify(delegate).getFreshTimestamp();
+        verifyLoggerInvokedOnceWithSpecificProfile("getFreshTimestamp", LONG_DURATION);
+        verifyLoggerInvokedOnceWithException(exception);
+    }
+
+    @Test
+    public void doesNotLogIfLockIsSlow() {
+        flushLogsWithCall(LONG_DURATION, () -> profilingTimelockService.lock(
+                LockRequest.of(ImmutableSet.of(StringLockDescriptor.of("exclusive")), 1234)));
+
+        verifyLoggerNeverInvoked();
+    }
+
+    @Test
+    public void doesNotLogIfWaitForLocksIsSlow() {
+        flushLogsWithCall(LONG_DURATION, () -> profilingTimelockService.waitForLocks(WaitForLocksRequest.of(
+                ImmutableSet.of(StringLockDescriptor.of("combination")), 123)));
+
+        verifyLoggerNeverInvoked();
+    }
+
+    @Test
+    public void logsSlowestOperationIfMultipleOperationsExceedTheSlowThresholdIfFirst() {
+        accumulateLogsWithCall(TWO_CENTURIES, profilingTimelockService::getFreshTimestamp);
+        flushLogsWithCall(TWO_CENTURIES, profilingTimelockService::startIdentifiedAtlasDbTransaction);
+
+        verifyLoggerInvokedOnceWithSpecificProfile("getFreshTimestamp", TWO_CENTURIES);
```

Comment:
still confused as to how you select `getFreshTimestamp`? is it just the first one?

Reply:
ugh, this test was badly copied. The second call is supposed to be `LONG_DURATION`.

For this case it'll always be the first (because of implementation details), but the intention was never to assert that